### PR TITLE
many pslint fixes

### PR DESCRIPTION
**Pull request by ShachafGoldstein**
##### SUMMARY

Handles:
* PSAvoidTrailingWhitespace
* PSAvoidGlobalVars
* PSAvoidAssignmentToAutomaticVariable
* PSAvoidUsingCmdletAliases
* PSAvoidUsingWriteHost
* PSUseDeclaredVarsMoreThanAssignments
* PSUsePSCredentialType
* PSAvoidUsingPositionalParameters
* PSAvoidUsingEmptyCatchBlock
* PSAvoidUsingWMICmdlet

[Original errors file](https://github.com/ansible/ansible/blob/devel/test/sanity/pslint/ignore.txt)

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
windows pslint errors

##### ADDITIONAL INFORMATION
More to come :smile:

+label: windows

### CODE REVIEW

File: lib/ansible/modules/windows/win_dsc.ps1 [hunk 1]
```diff
@@ -374,7 +374,9 @@ if ($resource.Module) {
 # engine updates the metadata propery. We don't care about any errors here
 try {
     Invoke-DscResource -Method Get -Property @{Fake="Fake"} @dsc_args > $null
-} catch {}
+} catch {
+    Out-Null;
```
Comment (jhawkesworth) [inline lib/ansible/modules/windows/win_dsc.ps1]: is trailing ; on the Out-Null needed?

Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/win_dsc.ps1]: No, in powershell you can leave out ; , but im used to writing with and i guess that one slipped

File: lib/ansible/modules/windows/win_domain.ps1 [hunk 2]
```diff
@@ -79,7 +79,7 @@ try {
 } catch [System.DirectoryServices.ActiveDirectory.ActiveDirectoryObjectNotFoundException] {
     Add-Warning -obj $result -message "Error during forest query initialization: $($_.Exception.Message)"
 } catch [System.DirectoryServices.ActiveDirectory.ActiveDirectoryOperationException] { i
```
Comment (jhawkesworth) [inline lib/ansible/modules/windows/win_domain.ps1]: any idea what the 'i' at the end of L81 is for? - it looks like it might be unintended to me.

Comment (jhawkesworth) [reply lib/ansible/modules/windows/win_domain.ps1]: I appreciate it was already there before!

File: lib/ansible/modules/windows/win_domain.ps1 [hunk 3]
```diff
@@ -17,6 +17,8 @@ Function Ensure-Prereqs {
         # NOTE: AD-Domain-Services includes: RSAT-AD-AdminCenter, RSAT-AD-Powershell and RSAT-ADDS-Tools
         $awf = Add-WindowsFeature AD-Domain-Services -WhatIf:$check_mode
         # FUTURE: Check if reboot necessary
+        $result.reboot_required = $awf.RestartNeeded
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_domain.ps1:19]: The changes in this module should be in a separate PR as they modify how the module behaves.

File: lib/ansible/modules/windows/win_domain_controller.ps1 [hunk 4]
```diff
@@ -24,6 +22,9 @@ Function Write-DebugLog {
 
     Write-Debug $msg
 
+    $log_path = $null
+    $log_path = Get-AnsibleParam -obj $params -name "log_path"
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_domain_controller.ps1]: This is a change to how the module behaves, should be in a separate PR.

File: lib/ansible/modules/windows/win_domain_controller.ps1 [hunk 5]
```diff
@@ -41,15 +42,16 @@ Function Get-MissingFeatures {
...
     return ,$missing_features # no, the comma's not a typo- allows us to return an empty array
 }
 
 Function Ensure-FeatureInstallation {
     # ensure RSAT-ADDS and AD-Domain-Services features are installed
 
-    Write-DebugLog "Ensuring required Windows features are installed..." 
+    Write-DebugLog "Ensuring required Windows features are installed..."
     $feature_result = Install-WindowsFeature $required_features
+    $result.reboot_required = $feature_result.RestartNeeded
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_domain_controller.ps1:52]: This is a change in how the module behaves, this should go into a separate PR.

File: lib/ansible/modules/windows/win_domain_controller.ps1 [hunk 6]
```diff
@@ -107,10 +109,8 @@ $read_only = Get-AnsibleParam -obj $params -name "read_only" -type "bool" -defau
 $site_name = Get-AnsibleParam -obj $params -name "site_name" -type "str" -failifempty $read_only
 
 $state = Get-AnsibleParam -obj $params -name "state" -validateset ("domain_controller", "member_server") -failifempty $result
-$log_path = Get-AnsibleParam -obj $params -name "log_path"
-$_ansible_check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -default $false
 
-$global:log_path = $log_path
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_domain_controller.ps1]: This looks like a change in how the module behaves. This should not be part of PSLint fixes.

File: lib/ansible/modules/windows/win_domain_membership.ps1 [hunk 7]
```diff
@@ -22,6 +20,9 @@ Function Write-DebugLog {
 
     Write-Debug $msg
 
+    $log_path = $null
+    $log_path = Get-AnsibleParam -obj $params -name "log_path"
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_domain_membership.ps1]: Please don't make changes to module behaviour in this PR.

File: lib/ansible/modules/windows/win_domain_membership.ps1 [hunk 8]
```diff
@@ -147,14 +148,14 @@ Function Set-Workgroup {
 
     Write-DebugLog ("Calling JoinDomainOrWorkgroup with workgroup {0}" -f $workgroup_name)
     try {
-        $swg_result = (Get-WmiObject -ClassName Win32_ComputerSystem).JoinDomainOrWorkgroup($workgroup_name)
+        $swg_result = Get-CimInstance Win32_ComputerSystem | Invoke-CimMethod -MethodName JoinDomainOrWorkgroup -Arguments @{Name="$workgroup_name"}
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_domain_membership.ps1:149]: Not sure if this should be part of this PR. May need to be discussed separately.

File: lib/ansible/modules/windows/win_find.ps1 [hunk 9]
```diff
@@ -303,7 +307,10 @@ Function Get-FilesInFolder($path) {
     try {
         $dir_files = $dir.EnumerateFileSystemInfos("*", [System.IO.SearchOption]::TopDirectoryOnly)
     } catch [System.IO.DirectoryNotFoundException] { # Broken ReparsePoint/Symlink, cannot enumerate
-    } catch [System.UnauthorizedAccessException] {}  # No ListDirectory permissions, Get-ChildItem ignored this
+    	  Add-Warning -obj $result -message "symlink enumeration error: $($_.Exception.Message)"
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_find.ps1]: Wrong indentation.

**Review by dagwieers — CHANGES_REQUESTED** [review]
I agree with 99% of the changes, however some of the changes modify the module's behaviour, or do not seem related to PSLint only. So I think we need to move those changes out of this PR to get this accepted.

File: lib/ansible/modules/windows/win_domain.ps1 (hunk 3 above)
Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/win_domain.ps1:19]: Honestly, I just added it on the fly since it seemed the most logical thing for that section, I don't mind removing it of course

File: lib/ansible/modules/windows/win_domain_controller.ps1 (hunk 4 above)
Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/win_domain_controller.ps1]: Why this is considered a behavior change?
 it just inserts the vars of that function into it.

File: lib/ansible/modules/windows/win_domain_controller.ps1 (hunk 5 above)
Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/win_domain_controller.ps1:52]: Same as with win_domain

File: lib/ansible/modules/windows/win_domain_controller.ps1 (hunk 6 above)
Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/win_domain_controller.ps1]: Same as with win_domain_controller

File: lib/ansible/modules/windows/win_domain_controller.ps1 (hunk 4 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_domain_controller.ps1]: Because before this wasn't done ?

File: lib/ansible/modules/windows/win_domain_membership.ps1 (hunk 8 above)
Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/win_domain_membership.ps1:149]: This was one of the problems with pslint and it is the same implementation just with CIM without any functional change

File: lib/ansible/modules/windows/win_domain.ps1 (hunk 3 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_domain.ps1:19]: It's not wrong, but it belongs in another PR where it will be discussed/accepted on its own merits.

File: lib/ansible/modules/windows/win_find.ps1 (hunk 9 above)
Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/win_find.ps1]: Fixed

File: lib/ansible/modules/windows/win_domain_membership.ps1 (hunk 8 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_domain_membership.ps1:149]: As I said, I am not sure. Someone else has to verify this.

File: lib/ansible/modules/windows/win_domain_controller.ps1 (hunk 4 above)
Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/win_domain_controller.ps1]: before, the same line was a global var outside the function that is why I don't think it should be considered a behavior change.

If this was a module it could have a lot of repercussions

File: lib/ansible/modules/windows/win_domain.ps1 (hunk 3 above)
Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/win_domain.ps1:19]: removed but for the record, I do think a different PR just for this is an overkill

**Comment by ShachafGoldstein** [comment]
* change | Out-Null to > $null
* Remove null assignments
* negotiate on .reboot_required = $awf.RestartNeeded

File: test/sanity/pslint/ignore.txt [hunk 10]
```diff
@@ -1,180 +1,106 @@
...
+test/integration/targets/win_script/files/test_script.ps1 PSAvoidUsingWriteHost # Keep
+test/integration/targets/win_script/files/test_script_with_args.ps1 PSAvoidUsingWriteHost # Keep
+test/integration/targets/win_script/files/test_script_with_splatting.ps1 PSAvoidUsingWriteHost # Keep
+lib/ansible/modules/windows/win_domain.ps1 PSAvoidUsingEmptyCatchBlock # Keep
+lib/ansible/modules/windows/win_dsc.ps1 PSAvoidUsingEmptyCatchBlock # Keep
+lib/ansible/modules/windows/win_find.ps1 PSAvoidUsingEmptyCatchBlock # Keep
+lib/ansible/modules/windows/win_region.ps1 PSAvoidUsingEmptyCatchBlock # Keep
+lib/ansible/modules/windows/win_uri.ps1 PSAvoidUsingEmptyCatchBlock # Keep
+lib/ansible/modules/windows/win_domain.ps1 PSUseDeclaredVarsMoreThanAssignments
+lib/ansible/modules/windows/win_domain_controller.ps1 PSUseDeclaredVarsMoreThanAssignments
```
Comment (mattclay) [inline test/sanity/pslint/ignore.txt]: This would be much easier to review if these entries remained on their original lines, since we want to make sure new entries are not being added.

Comment (ShachafGoldstein) [reply test/sanity/pslint/ignore.txt]: I moved them to the end to have a better order in the file for future work

**Comment by ShachafGoldstein** [comment]
When this one will be done, should i squash the commits to a single one again or leave it as it is?

**Comment by mattclay** [comment]
We'll squash as part of the merge, so leave it as-is to simplify reviews.

**Comment by ShachafGoldstein** [comment]
CI fails on "elapsed": 5.1011999999999995, " in a test that checks for ==5 ...

maybe fix that test in a different PR

File: lib/ansible/modules/windows/setup.ps1 [hunk 11]
```diff
@@ -422,14 +422,16 @@ if($gather_subset.Contains('winrm')) {
...
     }
 
     $winrm_cert_expiry = @()
     foreach ($winrm_cert_thumbprint in $winrm_cert_thumbprints) {
         Try {
-            $winrm_cert_expiry += Get-ChildItem -Path Cert:\LocalMachine\My | where Thumbprint -EQ $winrm_cert_thumbprint.Value.ToString().ToUpper() | select NotAfter
-        } Catch {}
+            $winrm_cert_expiry += Get-ChildItem -Path Cert:\LocalMachine\My | Where-Object Thumbprint -EQ $winrm_cert_thumbprint.Value.ToString().ToUpper() | Select-Object NotAfter
+        } Catch {
+            Add-Warning -obj $result -message "Error during certificate expiration retrieval"
```
Comment (jborean93) [inline lib/ansible/modules/windows/setup.ps1]: Should also add the exception message, e.g. `retrieval: $($_.Exception.Message)"`

File: lib/ansible/modules/windows/win_copy.ps1 [hunk 12]
```diff
@@ -34,7 +34,7 @@ $force = Get-AnsibleParam -obj $params -name "force" -type "bool" -default $true
 # used in query mode, contains the local files/directories/symlinks that are to be copied
 $files = Get-AnsibleParam -obj $params -name "files" -type "list"
 $directories = Get-AnsibleParam -obj $params -name "directories" -type "list"
-$symlinks = Get-AnsibleParam -obj $params -name "symlinks" -type "list"
+#$symlinks = Get-AnsibleParam -obj $params -name "symlinks" -type "list"
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_copy.ps1]: No need commented out unused code, just remove it altogether.

File: lib/ansible/modules/windows/win_credential.ps1 [hunk 13]
```diff
@@ -513,19 +513,19 @@ Function ConvertTo-CredentialAttribute {
 }
 
 Function Get-DiffInfo {
-    param($Credential)
+    param($Ansible_Credential)
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_credential.ps1]: If we really wanted to do this (and I don't think it is necessary), make it `AnsibleCredential` or `CredentialObj`.

File: lib/ansible/modules/windows/win_disk_facts.ps1 [hunk 14]
```diff
@@ -36,7 +36,7 @@ try {
 } catch {
     Fail-Json -obj $result -message "Failed to search the disks on the target: $($_.Exception.Message)"
 }
-[int32]$diskcount = $disks | Measure-Object | Select-Object  -ExpandProperty Count
+#[int32]$diskcount = $disks | Measure-Object | Select-Object  -ExpandProperty Count
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_disk_facts.ps1]: Remove instead of comment.

File: lib/ansible/modules/windows/win_domain.ps1 [hunk 15]
```diff
@@ -75,6 +76,7 @@ try {
     $forest_context = New-Object -TypeName System.DirectoryServices.ActiveDirectory.DirectoryContext -ArgumentList Forest, $dns_domain_name
     $forest = [System.DirectoryServices.ActiveDirectory.Forest]::GetForest($forest_context)
 } catch [System.DirectoryServices.ActiveDirectory.ActiveDirectoryObjectNotFoundException] {
+    Add-Warning -obj $result -message "Error during forest query initialization: $($_.Exception.Message)"
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_domain.ps1]: I wouldn't add this warning, it just means there wasn't a forest that was found which is expected in some situations.

File: lib/ansible/modules/windows/win_domain_membership.ps1 (hunk 7 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_domain_membership.ps1]: To add to this, prefer not to make this change, especially for an untested module. We can look at this in a separate PR if you wish.

File: lib/ansible/modules/windows/win_domain_controller.ps1 (hunk 4 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_domain_controller.ps1]: I agree this shouldn't be in this PR, keep behavioral changes to a separate PR.

File: lib/ansible/modules/windows/win_domain_membership.ps1 [hunk 16]
```diff
@@ -170,15 +170,15 @@ Function Join-Workgroup {
...
+            Remove-Computer -Workgroup $workgroup_name -Credential $domain_cred -Force
         } catch {
             Fail-Json -obj $result -message "failed to remove computer from domain: $($_.Exception.Message)"
         }
     }
 
     # we're already on a workgroup- change it.
     Else {
-        $swg_result = Set-Workgroup $workgroup_name
+        Set-Workgroup $workgroup_name
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_domain_membership.ps1:179]: Set-Workgroup looks like it returns something, we should case piping to `$null` or using the var.

File: lib/ansible/modules/windows/win_domain_membership.ps1 [hunk 17]
```diff
@@ -290,14 +286,14 @@ Try {
             If(-not $_ansible_check_mode) {
                 If(-not $workgroup_match) {
                     Write-DebugLog ("setting workgroup to {0}" -f $workgroup_name)
-                    $join_wg_result = Join-Workgroup -workgroup_name $workgroup_name -domain_admin_user $domain_admin_user -domain_admin_password $domain_admin_password
+                    Join-Workgroup -workgroup_name $workgroup_name -domain_admin_user $domain_admin_user -domain_admin_password $domain_admin_password
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_domain_membership.ps1:290]: `Join-Workgroup` can call `Set-Workgroup` which can return something, we will still need to make sure the output is put somewhere and not returned back to the controller.

File: lib/ansible/modules/windows/win_find.ps1 [hunk 18]
```diff
@@ -250,24 +250,28 @@ Function Get-FileStat($file) {
         try {
             $lnk_source = [Ansible.Command.SymLinkHelper]::GetSymbolicLinkTarget($file)
             $file_stat.lnk_source = $lnk_source
-        } catch {}
+        } catch {
+            #Add-Warning -obj $result -message "symlink source retrieval error: $($_.Exception.Message)"
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_find.ps1]: Best to keep this the way it was before instead of making a change that would affect the git blame.

File: lib/ansible/modules/windows/win_find.ps1 [hunk 19]
```diff
@@ -250,24 +250,28 @@ Function Get-FileStat($file) {
...
+
         # only get the size of a directory if there are files (not directories) inside the folder
         # Get-ChildItem -LiteralPath does not work properly on older OS', use .NET instead
         $dir_files = @()
         try {
             $dir_files = $file.EnumerateFiles("*", [System.IO.SearchOption]::AllDirectories)
         } catch [System.IO.DirectoryNotFoundException] { # Broken ReparsePoint/Symlink, cannot enumerate
-        } catch [System.UnauthorizedAccessException] {}  # No ListDirectory permissions, Get-ChildItem ignored this
-
+            Add-Warning -obj $result -message "symlink enumeration error: $($_.Exception.Message)"
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_find.ps1]: I would leave this as is for now. We need to look more into win_find and how it enumerates but prefer to do that sometime in the future without adding more warnings for now.

File: lib/ansible/modules/windows/win_find.ps1 [hunk 20]
```diff
@@ -250,24 +250,28 @@ Function Get-FileStat($file) {
...
+      	} catch [System.UnauthorizedAccessException] {  # No ListDirectory permissions, Get-ChildItem ignored this
+       	    Add-Warning -obj $result -message "symlink enumeration error: $($_.Exception.Message)"
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_find.ps1]: I would leave this as is for now. We need to look more into win_find and how it enumerates but prefer to do that sometime in the future without adding more warnings for now.I would leave this as is for now. We need to look more into win_find and how it enumerates but prefer to do that sometime in the future without adding more warnings for now.

File: lib/ansible/modules/windows/win_find.ps1 [hunk 21]
```diff
@@ -303,7 +307,10 @@ Function Get-FilesInFolder($path) {
     try {
         $dir_files = $dir.EnumerateFileSystemInfos("*", [System.IO.SearchOption]::TopDirectoryOnly)
     } catch [System.IO.DirectoryNotFoundException] { # Broken ReparsePoint/Symlink, cannot enumerate
-    } catch [System.UnauthorizedAccessException] {}  # No ListDirectory permissions, Get-ChildItem ignored this
+    	Add-Warning -obj $result -message "symlink enumeration error: $($_.Exception.Message)"
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_find.ps1]: I would leave this as is for now. We need to look more into win_find and how it enumerates but prefer to do that sometime in the future without adding more warnings for now.

File: lib/ansible/modules/windows/win_find.ps1 [hunk 22]
```diff
@@ -303,7 +307,10 @@ Function Get-FilesInFolder($path) {
...
+    } catch [System.UnauthorizedAccessException] {  # No ListDirectory permissions, Get-ChildItem ignored this
+        Add-Warning -obj $result -message "symlink enumeration error: $($_.Exception.Message)"
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_find.ps1]: I would leave this as is for now. We need to look more into win_find and how it enumerates but prefer to do that sometime in the future without adding more warnings for now.

File: lib/ansible/modules/windows/win_iis_website.ps1 [hunk 23]
```diff
@@ -18,14 +18,14 @@ $state = Get-AnsibleParam -obj $params -name "state" -type "str" -validateset "a
 $bind_port = Get-AnsibleParam -obj $params -name "port" -type "int"
 $bind_ip = Get-AnsibleParam -obj $params -name "ip" -type "str"
 $bind_hostname = Get-AnsibleParam -obj $params -name "hostname" -type "str"
-$bind_ssl = Get-AnsibleParam -obj $params -name "ssl" -type "str"
+#TODO: $bind_ssl = Get-AnsibleParam -obj $params -name "ssl" -type "str"
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_iis_website.ps1]: Remove if not used.

File: lib/ansible/modules/windows/win_pagefile.ps1 [hunk 24]
```diff
@@ -91,22 +90,20 @@ if ($state -eq "absent") {
     }
 
     # Make sure drive is accessible
-    if (($test_path) -and (-not (Test-Path "${drive}:"))) {
+    if (($testPath) -and (-not (Test-Path "${drive}:"))) {
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_pagefile.ps1]: This looks like it should be in it's own PR with a changelog and subsequently backported.

File: test/integration/targets/win_module_utils/library/camel_conversion_test.ps1 [hunk 25]
```diff
@@ -40,7 +40,7 @@ foreach ($entry in $output_dict.GetEnumerator()) {
     $key = $entry.Name
     $value = $entry.Value
 
-    $type = $value.GetType()
+    $value.GetType()
```
Comment (jborean93) [inline test/integration/targets/win_module_utils/library/camel_conversion_test.ps1]: If we aren't using `$type` just remove the call, we shouldn't be outputting anything to the output pipeline unless it's the return json.

**Review by jborean93 — CHANGES_REQUESTED** [review]
Everything mostly looks good, still a few things that should be removed and done in a separate PR. This is mostly the logging changes and some other behavioural stuff that I prefer not to make a change in a big PR like this.

File: lib/ansible/modules/windows/win_domain_membership.ps1 (hunk 17 above)
Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/win_domain_membership.ps1:290]: I can redirect to null

**Comment by ShachafGoldstein** [comment]
#57092 - For the log_path
#57093 - For win_pagefile bug

**Comment by ShachafGoldstein** [comment]
I remember now. I didn't send the output to null in the win_domain_* modules since they cause the CI to fail, the output is required @jborean93

**Comment by ShachafGoldstein** [comment]
@jborean93 @jhawkesworth - hopefully last review

File: lib/ansible/modules/windows/win_domain_user.ps1 [hunk 26]
```diff
@@ -185,6 +185,8 @@ If ($state -eq 'present') {
             } else {
                 $add_attributes.$attribute_name = $attribute_value
             }
+        } else {
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_domain_user.ps1]: I don't think this is right, I was not aware of `foreach` having an `else` statement in .NET.

File: lib/ansible/modules/windows/win_domain_membership.ps1 [hunk 27]
```diff
@@ -9,17 +9,19 @@ Set-StrictMode -Version 2
...
+$log_path = $null
+
 Function Write-DebugLog {
     Param(
-    [string]$msg
+        [string]$msg
     )
 
     $DebugPreference = "Continue"
+    $ErrorActionPreference = "Continue"
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_domain_membership.ps1]: Why was this added?

File: lib/ansible/modules/windows/win_domain_membership.ps1 [hunk 28]
```diff
@@ -168,15 +170,15 @@ Function Join-Workgroup {
...
+            Remove-Computer -Workgroup $workgroup_name -Credential $domain_cred -Force
         } catch {
             Fail-Json -obj $result -message "failed to remove computer from domain: $($_.Exception.Message)"
         }
     }
 
     # we're already on a workgroup- change it.
     Else {
-        $swg_result = Set-Workgroup $workgroup_name
+        Set-Workgroup $workgroup_name
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_domain_membership.ps1:179]: My comment still stands here, `Set-Workgroup` returns something and we should be redirecting it to `$null` to no confuse the json parser on the controller side.

File: lib/ansible/modules/windows/win_domain_membership.ps1 [hunk 29]
```diff
@@ -285,14 +289,14 @@ Try {
             If(-not $_ansible_check_mode) {
                 If(-not $workgroup_match) {
                     Write-DebugLog ("setting workgroup to {0}" -f $workgroup_name)
-                    $join_wg_result = Join-Workgroup -workgroup_name $workgroup_name -domain_admin_user $domain_admin_user -domain_admin_password $domain_admin_password
+                    Join-Workgroup -workgroup_name $workgroup_name -domain_admin_user $domain_admin_user -domain_admin_password $domain_admin_password
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_domain_membership.ps1:290]: `Join-Workgroup` can call `Set-Workgroup` which returns something, we need to redirect to `$null`.

**Review by jborean93 — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/windows/win_domain_user.ps1 (hunk 26 above)
Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/win_domain_user.ps1]: rebase fun - fixed

File: lib/ansible/modules/windows/win_domain_membership.ps1 (hunk 27 above)
Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/win_domain_membership.ps1]: no idea - removed

File: lib/ansible/modules/windows/win_domain_user.ps1 (hunk 26 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_domain_user.ps1]: aren't they always :)

File: lib/ansible/modules/windows/win_domain_membership.ps1 (hunk 28 above)
Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/win_domain_membership.ps1:179]: and my response too :), changing it breaks the CI - you want to keep it or should I look into changing the tests?

File: lib/ansible/modules/windows/win_domain_membership.ps1 (hunk 29 above)
Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/win_domain_membership.ps1:290]: changing it breaks the CI - you want to keep it or should I look into changing the tests?

File: lib/ansible/modules/windows/win_domain_membership.ps1 (hunk 28 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_domain_membership.ps1:179]: Hmm I don't see how based on looking at the code, do you have a link to the failure? I'm just looking at `Set-Workgroup` and the return statement is on a newline but still part of the failing brace so it doesn't even look like that does anything https://github.com/ansible/ansible/blob/66e0041b3ab5d24ae51f28938cf6a90af60374e6/lib/ansible/modules/windows/win_domain_membership.ps1#L158. So maybe just return that line from the function as the module will exit out when calling `Fail-Json`.

Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/win_domain_membership.ps1:179]: https://app.shippable.com/github/ansible/ansible/runs/125201

Comment (jborean93) [reply lib/ansible/modules/windows/win_domain_membership.ps1:179]: Ahhh, it's because we call `Fail-Json` which outputs the return json in the output stream instead of stdout. While we exit it early the json string is redirected. Ok If you can just fix up the indentation of line 158 and remove that return statement I think we are all good.

Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/win_domain_membership.ps1:179]: you want the function to not return anything then?

Comment (jborean93) [reply lib/ansible/modules/windows/win_domain_membership.ps1:179]: Yes, the `Set-Workgroup` function has the following line https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/windows/win_domain_membership.ps1#L156. It makes it look like the function returns that result but because the brace for the if statement is placed after this call it is actually being run after `Fail-Json`. `Fail-Json` stops the pipeline so nothing will ever be called making the `return` line unreachable. So right now it works it's just horribly confusing and leads to the conversation above.

Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/win_domain_membership.ps1:179]: removed, let's see what happens - fingers crossed

File: lib/ansible/modules/windows/win_domain_membership.ps1 [hunk 30]
```diff
@@ -168,15 +168,15 @@ Function Join-Workgroup {
...
+            Remove-Computer -Workgroup $workgroup_name -Credential $domain_cred -Force
         } catch {
             Fail-Json -obj $result -message "failed to remove computer from domain: $($_.Exception.Message)"
         }
     }
 
     # we're already on a workgroup- change it.
     Else {
-        $swg_result = Set-Workgroup $workgroup_name
+        Set-Workgroup $workgroup_name > $null
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_domain_membership.ps1]: So now that we've cleared up the confusion and found that the return statement in `Set-Workgroup` isn't needed you can remove the `> $null` part from here.

File: lib/ansible/modules/windows/win_domain_membership.ps1 [hunk 31]
```diff
@@ -285,14 +287,14 @@ Try {
             If(-not $_ansible_check_mode) {
                 If(-not $workgroup_match) {
                     Write-DebugLog ("setting workgroup to {0}" -f $workgroup_name)
-                    $join_wg_result = Join-Workgroup -workgroup_name $workgroup_name -domain_admin_user $domain_admin_user -domain_admin_password $domain_admin_password
+                    Join-Workgroup -workgroup_name $workgroup_name -domain_admin_user $domain_admin_user -domain_admin_password $domain_admin_password > $null
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_domain_membership.ps1]: Same here, no need for `> $null` anymore.

**Comment by ShachafGoldstein** [comment]
win_domain_membership still fails with `> $null` - i.e https://app.shippable.com/github/ansible/ansible/runs/130900/42/tests

File: lib/ansible/modules/windows/win_domain_membership.ps1 (hunk 30 above)
Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/win_domain_membership.ps1]: done

File: lib/ansible/modules/windows/win_domain_membership.ps1 (hunk 31 above)
Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/win_domain_membership.ps1]: done

**Review by jborean93 — APPROVED** [review]
Ok, looks like CI is green and this is as much as I want to change in 1 PR.
